core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -186,30 +202,63 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
-    TxnTransitMetadata(producerId, producerEpoch, txnTimeoutMs, state, topicPartitions.toSet, txnStartTimestamp, txnLastUpdateTimestamp)
+    TxnTransitMetadata(producerId, lastProducerId, producerEpoch, lastProducerEpoch, txnTimeoutMs, state, topicPartitions.toSet,
+      txnStartTimestamp, txnLastUpdateTimestamp)
   }
 
   def prepareFenceProducerEpoch(): TxnTransitMetadata = {
     if (producerEpoch == Short.MaxValue)
       throw new IllegalStateException(s"Cannot fence producer with epoch equal to Short.MaxValue since this would overflow")
 
-    prepareTransitionTo(PrepareEpochFence, producerId, (producerEpoch + 1).toShort, txnTimeoutMs, topicPartitions.toSet,
-      txnStartTimestamp, txnLastUpdateTimestamp)
+    prepareTransitionTo(PrepareEpochFence, producerId, (producerEpoch + 1).toShort, RecordBatch.NO_PRODUCER_EPOCH, txnTimeoutMs,
+      topicPartitions.toSet, txnStartTimestamp, txnLastUpdateTimestamp)
   }
 
-  def prepareIncrementProducerEpoch(newTxnTimeoutMs: Int, updateTimestamp: Long): TxnTransitMetadata = {
+  def prepareIncrementProducerEpoch(newTxnTimeoutMs: Int,
+                                    currentProducerEpoch: Option[Short],
+                                    updateTimestamp: Long): Either[Errors, TxnTransitMetadata] = {
     if (isProducerEpochExhausted)
       throw new IllegalStateException(s"Cannot allocate any more producer epochs for producerId $producerId")
 
-    val nextEpoch = if (producerEpoch == RecordBatch.NO_PRODUCER_EPOCH) 0 else producerEpoch + 1
-    prepareTransitionTo(Empty, producerId, nextEpoch.toShort, newTxnTimeoutMs, immutable.Set.empty[TopicPartition], -1,
-      updateTimestamp)
+    // If no epoch was provided by the producer, bump the current epoch and set the last epoch to -1
+    // If an epoch was provided by the producer:
+    //   If it matches the current epoch, the producer is attempting to continue after an error, and no other producer
+    // has been initialized. Bump the current and last epochs.
+    //   If it matches the previous epoch, it is a retry of a successful call, so just return the current epoch without
+    // bumping. There is no danger of this producer being fenced, because a new producer calling InitProducerId would
+    // have caused the last epoch to be set to -1.
+    //   Otherwise, the producer has a stale epoch and should receive an INVALID_PRODUCER_ERROR
+    val epochBumpResult: Either[Errors, (Int, Short)] = currentProducerEpoch match {
+      case None =>
+        Right(if (producerEpoch == RecordBatch.NO_PRODUCER_EPOCH) 0 else producerEpoch + 1, RecordBatch.NO_PRODUCER_EPOCH)
+
+      case Some(currentEpoch) =>
+        if (currentEpoch == producerEpoch)
+          Right(producerEpoch + 1, producerEpoch)
+        else if (currentEpoch == lastProducerEpoch)
+          Right(producerEpoch, lastProducerEpoch)
+        else {
+          error(s"Provided producer epoch $currentEpoch does not match current " +
```

Comment:
I am not sure if `error` is appropriate. There's no problem from the perspective of the coordinator. We have just encountered a fenced client. Perhaps `info` would be ok?